user.go

```diff
@@ -0,0 +1,74 @@
+package goldap
+
+import (
+	"fmt"
+
+	"github.com/go-ldap/ldap/v3"
+)
+
+// ReadUser reads ldap user and return it's attributes on an error if the user donesn't exist
+func (c *Client) ReadUser(ou string, name string, sAMAccountName string, upn string) (entries map[string][]string, err error) {
+	filter := ""
+
+	if name != "" {
+		filter = fmt.Sprintf("%s(name=%s)", filter, name)
+	}
+
+	if sAMAccountName != "" {
+		filter = fmt.Sprintf("%s(sAMAccountName=%s)", filter, sAMAccountName)
+	}
+
+	if upn != "" {
+		filter = fmt.Sprintf("%s(userPrincipalName=%s)", filter, upn)
+	}
+
+	filter = fmt.Sprintf("(&(objectCategory=person)(objectClass=user)%s)", filter)
```

Comment:
At first read I was a bit confused by how the code build the filter, I suggest instead to use an array to store filter "conditions" and join them after. Something like that:
```
conditions := []string{"(objectCategory=person)", "(objectClass=user)"}

if name != "" {
	conditions = append(conditions, fmt.Sprintf("(name=%s)", name))
}

....

filter := fmt.Sprintf("(&%s)", strings.Join(conditions, ""))
```
What do you think?

Reply:
Great. Your suggestion sounds good to me.

For the records, I'm not a go expert ^^

Reply:
it's much better ^^

Reply:
Great :)